### PR TITLE
[macos10.13-beta1] FinderSync and MediaLibrary bindings

### PR DESCRIPTION
**Pull request by timrisi**

### CODE REVIEW

File: src/findersync.cs [hunk 1]
```diff
@@ -27,6 +27,24 @@ interface FIFinderSyncController : NSSecureCoding, NSCopying
 
 		[NullAllowed, Export ("selectedItemURLs")]
 		NSUrl[] SelectedItemURLs { get; }
+
+		[Mac (10,13, onlyOn64 : true)]
+		[Export ("lastUsedDateForItemWithURL:")]
+		[return: NullAllowed]
+		NSDate GetLastUsedDateForItem (NSUrl itemURL);
```
Comment (rolfbjarne) [inline src/findersync.cs]: IMHO this should be `GetLastUsedDate` (to match `SetLastUsedDate`)

File: src/findersync.cs [hunk 2]
```diff
@@ -27,6 +27,24 @@ interface FIFinderSyncController : NSSecureCoding, NSCopying
...
+
+		[Mac (10,13, onlyOn64 : true)]
+		[Export ("setLastUsedDate:forItemWithURL:completion:")]
+		void SetLastUsedDate (NSDate lastUsedDate, NSUrl itemURL, Action<NSError> completion);
+
+		[Mac (10,13, onlyOn64 : true)]
+		[Export ("tagDataForItemWithURL:")]
+		[return: NullAllowed]
+		NSData GetTagDataForItem (NSUrl itemURL);
```
Comment (rolfbjarne) [inline src/findersync.cs]: Same: `GetTagData`

File: src/findersync.cs (hunk 1 above)
Comment (rolfbjarne) [inline src/findersync.cs]: Also managed code uses `Url`, not `URL`

File: src/findersync.cs (hunk 2 above)
Comment (rolfbjarne) [inline src/findersync.cs]: Same: `itemUrl` instead of `itemURL`.

File: src/findersync.cs [hunk 3]
```diff
@@ -27,6 +27,24 @@ interface FIFinderSyncController : NSSecureCoding, NSCopying
...
 		NSUrl[] SelectedItemURLs { get; }
+
+		[Mac (10,13, onlyOn64 : true)]
+		[Export ("lastUsedDateForItemWithURL:")]
+		[return: NullAllowed]
+		NSDate GetLastUsedDateForItem (NSUrl itemURL);
+
+		[Mac (10,13, onlyOn64 : true)]
+		[Export ("setLastUsedDate:forItemWithURL:completion:")]
+		void SetLastUsedDate (NSDate lastUsedDate, NSUrl itemURL, Action<NSError> completion);
```
Comment (rolfbjarne) [inline src/findersync.cs]: Same: `itemUrl`.

File: src/findersync.cs [hunk 4]
```diff
@@ -54,6 +72,21 @@ interface FIFinderSyncProtocol
 
 		[Export ("toolbarItemToolTip")]
 		string ToolbarItemToolTip { get; }
+
+	    [Export ("supportedMessageInterfaceNamesForItemWithURL:")]
```
Comment (rolfbjarne) [inline src/findersync.cs]: This has space indentation instead of tab indentation.

File: src/findersync.cs [hunk 5]
```diff
@@ -54,6 +72,21 @@ interface FIFinderSyncProtocol
...
+	    string[] GetSupportedMessageInterfaceNamesForItem (NSUrl itemURL);
```
Comment (rolfbjarne) [inline src/findersync.cs]: Same: `itemUrl`

Also I think `GetSupportedMessageInterfaceNames` would be simpler and still won't lose information.

File: src/findersync.cs [hunk 6]
```diff
@@ -54,6 +72,21 @@ interface FIFinderSyncProtocol
...
+
+		[Mac (10,13, onlyOn64 : true)]
+		[Export ("protocolForMessageInterface:")]
+		Protocol GetProtocol (NSFileProviderMessageInterface messageInterface);
+
+		[Mac (10,13, onlyOn64 : true)]
+		[Export ("exportedObjectForMessageInterface:itemURL:error:")]
+		NSObject GetExportedObject (NSFileProviderMessageInterface messageInterface, NSUrl itemURL, [NullAllowed] out NSError error);
```
Comment (rolfbjarne) [inline src/findersync.cs]: Same: `itemUrl`.

File: src/findersync.cs [hunk 7]
```diff
@@ -54,6 +72,21 @@ interface FIFinderSyncProtocol
...
+
+		[Mac (10,13, onlyOn64 : true)]
+		[Export ("valuesForAttributes:forItemWithURL:completion:")]
+		void ValuesForAttributes (string[] attributes, NSUrl itemURL, Action<NSDictionary<NSString, NSObject>, NSError> completion);
```
Comment (rolfbjarne) [inline src/findersync.cs]: Just `GetValues` would be simpler.

Also: `itemUrl`.

File: src/medialibrary.cs [hunk 8]
```diff
@@ -156,6 +156,18 @@ interface MediaLibraryTypeIdentifierKey {
...
+		[Field ("MLPhotosLivePhotosGroupTypeIdentifier")]
+		NSString MLPhotosLivePhotosGroupTypeIdentifier { get; }
+
+		[Mac (10,13, onlyOn64 : true)]
+		[Field ("MLPhotosLongExposureGroupTypeIdentifier")]
+		NSString MLPhotosLongExposureGroupTypeIdentifier { get; }
+
+		[Mac (10,13, onlyOn64 : true)]
+		[Field ("MLPhotosAnimatedGroupTypeIdentifier")]
+		NSString MLPhotosAnimatedGroupTypeIdentifier { get; }
```
Comment (rolfbjarne) [inline src/medialibrary.cs]: Remove the `ML` prefix for all three constants.

File: src/findersync.cs [hunk 9]
```diff
@@ -27,6 +27,24 @@ interface FIFinderSyncController : NSSecureCoding, NSCopying
...
 		NSUrl[] SelectedItemURLs { get; }
+
+		[Mac (10,13, onlyOn64 : true)]
+		[Export ("lastUsedDateForItemWithURL:")]
+		[return: NullAllowed]
+		NSDate GetLastUsedDate (NSUrl itemUrl);
+
+		[Mac (10,13, onlyOn64 : true)]
+		[Export ("setLastUsedDate:forItemWithURL:completion:")]
+		void SetLastUsedDate (NSDate lastUsedDate, NSUrl itemUrl, Action<NSError> completion);
```
Comment (spouliot) [inline src/findersync.cs]: Async ?

File: src/findersync.cs [hunk 10]
```diff
@@ -27,6 +27,24 @@ interface FIFinderSyncController : NSSecureCoding, NSCopying
...
+
+		[Mac (10,13, onlyOn64 : true)]
+		[Export ("tagDataForItemWithURL:")]
+		[return: NullAllowed]
+		NSData GetTagData (NSUrl itemUrl);
+
+		[Mac (10,13, onlyOn64 : true)]
+		[Export ("setTagData:forItemWithURL:completion:")]
+		void SetTagData ([NullAllowed] NSData tagData, NSUrl itemUrl, Action<NSError> completion);
```
Comment (spouliot) [inline src/findersync.cs]: A sync ?

File: src/findersync.cs [hunk 11]
```diff
@@ -54,6 +72,21 @@ interface FIFinderSyncProtocol
 
 		[Export ("toolbarItemToolTip")]
 		string ToolbarItemToolTip { get; }
+
+		[Export ("supportedMessageInterfaceNamesForItemWithURL:")]
+		string[] GetSupportedMessageInterfaceNames (NSUrl itemURL);
```
Comment (spouliot) [inline src/findersync.cs]: Url, not URL

File: src/findersync.cs [hunk 12]
```diff
@@ -54,6 +72,21 @@ interface FIFinderSyncProtocol
...
+		Protocol GetProtocol (NSFileProviderMessageInterface messageInterface);
+
+		[Mac (10,13, onlyOn64 : true)]
+		[Export ("exportedObjectForMessageInterface:itemURL:error:")]
+		NSObject GetExportedObject (NSFileProviderMessageInterface messageInterface, NSUrl itemUrl, [NullAllowed] out NSError error);
+
+		[Mac (10,13, onlyOn64 : true)]
+		[Export ("valuesForAttributes:forItemWithURL:completion:")]
+		void GetValues (string[] attributes, NSUrl itemUrl, Action<NSDictionary<NSString, NSObject>, NSError> completion);
 	}
```
Comment (spouliot) [inline src/findersync.cs]: Action too complex - we can guess the usage
Better use a custom delegate type in such case, it's self documenting

File: src/findersync.cs [hunk 13]
```diff
@@ -83,11 +83,12 @@ interface FIFinderSyncProtocol
 		[Export ("supportedServiceNamesForItemWithURL:")]
 		string[] SupportedServiceNames (NSUrl itemUrl);
 
+#if FALSE // TODO: Activate after 10.13 foundation APIs have been merged
```
Comment (rolfbjarne) [inline src/findersync.cs]: Can you file a bug about this so that we don't forget?

Comment (timrisi) [inline src/findersync.cs]: Whoops. I did file a bug, meant to add it to the comment. I'll update it with the bug #

**Comment by timrisi** [comment]
Known failure https://bugzilla.xamarin.com/show_bug.cgi?id=57762

**Pull request merged**